com.unity.render-pipelines.core/Runtime/Inputs/InputRegistering.cs

```diff
@@ -76,18 +135,39 @@ public static void RegisterInputs(List<InputManagerEntry> entries)
 
             var inputManager = assets[0];
 
-            // Wrap in serialized object
+            // Wrap in serialized object to access c++ fields
             var soInputManager = new SerializedObject(inputManager);
             var spAxes = soInputManager.FindProperty("m_Axes");
 
-            foreach (InputManagerEntry entry in entries)
-            {
-                WriteEntry(spAxes, entry);
-            }
+            // At this point, we assume that entries in spAxes are already unique.
+
+            // Ensure no double entry are tried to be registered (trim early)
+            MakeUniquePendingInputsToRegister();
+
+            // Cache already existing entries to minimaly use serialization
+            var cachedEntries = GetCachedInputs(spAxes);
+            // And trim pending entries regarding already cached ones.
+            RemovePendingInputsToAddThatAreAlreadyRegistered(cachedEntries, s_PendingInputsToRegister);
+
+            // Add now unique entries
+            AddEntriesWithoutCheck(spAxes, s_PendingInputsToRegister);
 
             // Commit
             soInputManager.ApplyModifiedProperties();
         }
+
+        public static void RegisterInputs(List<InputManagerEntry> entries)
+        {
+#if ENABLE_INPUT_SYSTEM && ENABLE_INPUT_SYSTEM_PACKAGE
+            Debug.LogWarning("Trying to add entry in the legacy InputManager but using InputSystem package. Skiping.");
+            return;
+#else
+            s_PendingInputsToRegister.AddRange(entries);
+
+            //delay the call in order to do only one pass event if several different class register inputs
+            EditorApplication.delayCall += DelayedRegisterInput;
```

Comment:
does this mean we could end up calling this function multiple times in the same frame?
also when do we unregister the call?

Reply:
No it means you can call it multiple time in a frame, it will all be batched in delayed call.
Example:
Frame1, two different classes call register input
```
RegisterInputs(new List<InputManagerEntry>
      {
            new InputManagerEntry { name = "cmd1",  kind = InputManagerEntry.Kind.KeyOrButton, btnPositive = "left ctrl",   altBtnPositive = "joystick button 8" },
            new InputManagerEntry { name = "cmd2",  kind = InputManagerEntry.Kind.KeyOrButton, btnPositive = "backspace",   altBtnPositive = "joystick button 9" }
      });
      
      
RegisterInputs(new List<InputManagerEntry>
      {
            new InputManagerEntry { name = "cmd3",  kind = InputManagerEntry.Kind.KeyOrButton, btnPositive = "left alt",   altBtnPositive = "joystick button 1" }
      });
```
Delayed  calls:
- register cmd1, cmd2 and cmd3 at the same time
- nothing to register, so exit early

So we inspected the already registered inputs only once instead of two time when done at same frame.



And for unregistering, the mechanism where not here prior and was never requested. I assume it is not needed in this case. But it can be implemented in a similar way. I assume this was done to add inputs in the InputManager.asset only and not remove them.